File: flutter/lib/presentation/components/bars/flashcard_app_bar.dart
import 'package:flashcard/core/utils/navigation_helper.dart';
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

class FlashcardAppBar extends StatelessWidget implements PreferredSizeWidget {
  final Widget? leadingWidget;
  final String? title;
  final Widget? actions;
  final double height;
  const FlashcardAppBar({super.key, this.leadingWidget, this.title, this.actions, this.height=kToolbarHeight});

  String _setDefaultTitle(String route){
    switch (route) {
      case '/':
        return 'Home';
      case '/deck':
        return 'Your Decks';
      case '/settings':
        return 'Settings';
      default:
        return 'Flashcard AI';
    }
  }

  @override
  Widget build(BuildContext context) {
    final titleText = title??_setDefaultTitle(getCurrentRouteName(context));
    final double topPadding = MediaQuery.of(context).padding.top;
    return SizedBox(
      height: preferredSize.height + topPadding + 24,
      child: Padding(padding: EdgeInsets.only(top: topPadding) ,
          child: Stack(
            children: [
              Center(child: Text(titleText, style: TextStyle(color: Colors.black87),)),
              SizedBox(
                height: height,
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      (leadingWidget == null)? (context.canPop())?
                        IconButton(
                          icon: const Icon(Icons.arrow_back, color: Colors.black87),
                          onPressed: () {
                            // Navigate back to the previous screen
                            context.pop();
                          },
                        )
                          : const SizedBox(width: 16.0, height: 16.0,)
                    : leadingWidget!,
                    actions != null? leadingWidget!
                        : const SizedBox(width: 16.0, height: 16.0,)
                  ],
                ),
              ),
            ],
          )),
    );
  }

  @override
  Size get preferredSize => Size.fromHeight(height);


}

File: android/app/src/main/java/com/synergyboat/flashcardAi/presentation/components/FlashcardAppBar.kt
package com.synergyboat.flashcardAi.presentation.components

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.*
import androidx.core.view.WindowInsetsCompat
import androidx.navigation.NavController

@Composable
fun FlashcardAppBar(
    navController: NavController,
    title: String? = null,
    leadingContent: (@Composable (() -> Unit))? = null,
    actions: (@Composable RowScope.() -> Unit)? = null,
    height: Dp = 56.dp,
) {
    val route = navController.currentDestination?.route ?: "/"
    val defaultTitle = when (route) {
        "/" -> "Home"
        "deck" -> "Your Decks"
        "settings" -> "Settings"
        else -> "Flashcard AI"
    }

    val resolvedTitle = title ?: defaultTitle
    val view = LocalView.current
    val topPadding = with(LocalDensity.current) {
        WindowInsetsCompat.toWindowInsetsCompat(view.rootWindowInsets)
            .getInsets(WindowInsetsCompat.Type.statusBars()).top.toDp()
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .height(height + topPadding + 24.dp)
            .padding(top = topPadding)
    ) {
        Box(modifier = Modifier.fillMaxSize()) {
            // Center title
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .align(Alignment.Center),
                horizontalArrangement = Arrangement.Center
            ) {
                Text(
                    text = resolvedTitle,
                    style = MaterialTheme.typography.titleMedium.copy(
                        color = Color.Black,
                        fontWeight = FontWeight.Medium
                    )
                )
            }

            // Leading & actions
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(height),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                if (leadingContent != null) {
                    leadingContent()
                } else if (navController.previousBackStackEntry != null) {
                    IconButton(onClick = { navController.popBackStack() }) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Back", tint = Color.Black)
                    }
                } else {
                    Spacer(modifier = Modifier.width(16.dp))
                }

                if (actions != null) {
                    Row(content = actions)
                } else {
                    Spacer(modifier = Modifier.width(16.dp))
                }
            }
        }
    }
}